localization/ndt_scan_matcher/src/pose_array_interpolator.cpp

```diff
@@ -0,0 +1,105 @@
+// Copyright 2015-2019 Autoware Foundation
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "ndt_scan_matcher/pose_array_interpolator.hpp"
+
+PoseArrayInterpolator::PoseArrayInterpolator(
+  rclcpp::Node * node, const rclcpp::Time target_ros_time,
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array)
+: logger_(node->get_logger()),
+  clock_(*node->get_clock()),
+  current_pose_ptr_(new PoseWithCovarianceStamped),
+  old_pose_ptr_(new PoseWithCovarianceStamped),
+  new_pose_ptr_(new PoseWithCovarianceStamped),
+  success_(true)
+{
+  get_nearest_timestamp_pose(pose_msg_ptr_array, target_ros_time, old_pose_ptr_, new_pose_ptr_);
+  const geometry_msgs::msg::PoseStamped interpolated_pose_msg =
+    interpolate_pose(*old_pose_ptr_, *new_pose_ptr_, target_ros_time);
+  current_pose_ptr_->header = interpolated_pose_msg.header;
+  current_pose_ptr_->pose.pose = interpolated_pose_msg.pose;
+  current_pose_ptr_->pose.covariance = old_pose_ptr_->pose.covariance;
+}
+
+PoseArrayInterpolator::PoseArrayInterpolator(
+  rclcpp::Node * node, const rclcpp::Time target_ros_time,
+  const std::deque<PoseWithCovarianceStamped::ConstSharedPtr> & pose_msg_ptr_array,
+  const double & pose_timeout_sec, const double & pose_distance_tolerance_meters)
+: PoseArrayInterpolator(node, target_ros_time, pose_msg_ptr_array)
+{
+  // check the time stamp
+  success_ =
```

Comment:
The variable `success_` seems to be overwritten. 
Maybe it is necessary to take OR of them?

Reply:
You are right... thanks for pointing it out.
https://github.com/autowarefoundation/autoware.universe/pull/1757/commits/f80b03f9f473fcd64725e675191da85ce4720d4d (took AND for each of them)

Reply:
I can also suggest the one below. 
This way can reduce the state change of `success_`. 

```cpp
  const bool b1 = validate_time_stamp_difference(
                          old_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);
  const bool b2 = validate_time_stamp_difference(
                          new_pose_ptr_->header.stamp, target_ros_time, pose_timeout_sec);

  // check the position jumping (ex. immediately after the initial pose estimation)
  const bool b3 = validate_position_difference(
                          old_pose_ptr_->pose.pose.position, new_pose_ptr_->pose.pose.position,
                          pose_distance_tolerance_meters);

  success_ = b1 && b2 && b3;
```

Reply:
Thanks again.
Resolved :+1: 
https://github.com/autowarefoundation/autoware.universe/pull/1757/commits/a41b7c47bd16c15b0dbd44a8129d1080f568def4